### PR TITLE
Allow chaining .else()

### PR DESCRIPTION
**Pull request by minamorl**
Related to #3.
@wavebeem Could you review on this?

### CODE REVIEW

File: src/monoflow.ts [hunk 1]
```diff
@@ -17,47 +18,58 @@ export class Workflow<Z, A, B> {
...
+      console.log("here");
+      const binded = this._err.bind(this);
+      const safeFn = (err: Error) => {
+        try {
+          return binded(err);
+        } catch (err2) {
+          return err2;
+        }
+      }
+      return new Workflow(undefined, (error: Error) => fn(safeFn(error) as Error), [...this._steps]);
```
Comment (wavebeem) [inline src/monoflow.ts]: This seems incorrect to me. `fn` should only be called with `Error`, not `B`

Comment (wavebeem) [reply src/monoflow.ts]:
```ts
  const workflow = Workflow.create((n: number) => {
    throw new Error("hey");
  })
    .else((_err) => {
      return "hey";
    })
    .else((err) => {
      // oops! `err` is actually a string, not Error type :(
      return `${err.message}!`;
    });
  expect(workflow.run(1)).toBe("hey!");
```

File: src/monoflow.ts [hunk 2]
```diff
@@ -17,47 +18,58 @@ export class Workflow<Z, A, B> {
...
+        const nextSteps: Workflow<any, any, any> | undefined = iter.next().value;
+        if (!nextSteps?._err) {
+          throw err as Error;
         }
-        // TODO: Should you be able to chain `.else` repeatedly? I don't think
-        // this will work if you do that. This would probably need to be
-        // recursive.
+        ret = nextSteps._err(err as Error);
       }
-      i++;
```
Comment (wavebeem) [inline src/monoflow.ts:60]: I prefer to not use iterators manually, integers are easier imo.

**Review by wavebeem — COMMENTED** [review]
There's a problem with `else` behavior

I also think that the logic should be in `run`. `then` and `else` build the data structure, and `run` interprets it based on the input. It seems like the correct spot to make decisions.

File: src/monoflow.ts (hunk 1 above)
Comment (minamorl) [reply src/monoflow.ts]: OK, I fixed to ignore the second one if the first one returns value correctly.

File: src/monoflow.ts (hunk 2 above)
Comment (minamorl) [reply src/monoflow.ts:60]: Another PR?

**Comment by minamorl** [comment]
Thanks for the review. There's some thoughts on your point.

1. I think it's better memory efficiency. Because current implementation creates bigger workflow each time invoking `then()`. I don't think it's good idea.
2. Writing every logic on `.run` is huge i guess? It sounds tricky.
3. My implementations of `.then` and `.else` just trying to determine if it's continuous. That's simple. 

Basically, I don't care how it is implemented as, while it works well. This PR is just for fixing the behavior of  `.else()` chain.
I think if you have better idea, making an issue is good to start? :smile:

File: src/monoflow.ts (hunk 2 above)
Comment (wavebeem) [reply src/monoflow.ts:60]: Yeah, it's not a big deal

File: src/monoflow.ts [hunk 3]
```diff
@@ -17,47 +18,63 @@ export class Workflow<Z, A, B> {
...
+      const safeFn = (err: Error) => {
+        try {
+          return binded(err);
+        } catch (err2) {
+          return err2;
+        }
+      }
+      return new Workflow(undefined, (error: Error) => {
+        const result = safeFn(error);
+        if (result instanceof Error) {
```
Comment (wavebeem) [inline src/monoflow.ts]: This won't work if you throw a non-Error type:

`throw { message: "this is bad" }`

Comment (minamorl) [reply src/monoflow.ts]: Thanks. I completely forget about it :+1:

File: src/monoflow.ts [hunk 4]
```diff
@@ -29,21 +28,28 @@ export class Workflow<Z, A, B> {
 
   else<C>(fn: (error: Error) => C): Workflow<Z, B, unknown>  {
     // Make a higher-order function when `this._err` exists.
+    const sym = Symbol.for("monoflow.hasError");
```
Comment (wavebeem) [inline src/monoflow.ts]: You could simplify this by using a string. Monoflow users won't see this object, so you don't need to hide the keys.

**Review by wavebeem — APPROVED** [review]

File: src/monoflow.ts (hunk 4 above)
Comment (minamorl) [reply src/monoflow.ts]: that's true! fixed.

**Pull request merged**